File: c/compiler/step1_read_print.c
#include <editline/readline.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "printer.h"
#include "reader.h"

MalType* READ(char *str) {
  return read_str(str);
}

MalType* EVAL(MalType *ast) {
  return ast;
}

char* PRINT(MalType *ast) {
  return pr_str(ast, 2);
}

char* rep(char *str) {
  return PRINT(EVAL(READ(str)));
}

int main() {
  char *buffer;

  read_history("history.txt");
  while ((buffer = readline("user> ")) != NULL) {
    printf("%s\n", rep(buffer));
    if (strlen(buffer) > 0) {
      add_history(buffer);
    }
    free(buffer);
  }
  write_history("history.txt");

  return 0;
}

File: impls/frust/step1_read_print.rs
const HISTORY: &str = "history.txt";

mod common;
mod printer;
mod reader;
mod types;

pub use common::*;

fn main() {
    let rl_config = rustyline::Config::builder()
        .auto_add_history(true)
        .completion_type(rustyline::CompletionType::Circular)
        .edit_mode(rustyline::EditMode::Vi)
        .color_mode(rustyline::ColorMode::Enabled)
        .build();
    let mut rl = rustyline::Editor::<()>::with_config(rl_config);
    let _ = rl.load_history(HISTORY);
    for readline in rl.iter("user> ") {
        match readline {
            Ok(s) => {
                match rep(s) {
                    Ok(s) => println!("{}", s),
                    Err(e) => eprintln!("{}", e),
                };
            }
            Err(e) => eprintln!("{}", e),
        }
    }
    rl.save_history(HISTORY).expect("Failed to save history");
}

fn rep(s: String) -> Result<String> {
    let r = read(s)?;
    let e = eval(r)?;
    Ok(print(e))
}

fn read(s: String) -> MalResult {
    reader::read_str(s)
}

fn eval(t: types::MalType) -> MalResult {
    Ok(t)
}

fn print(t: types::MalType) -> String {
    printer::pr_str(&t, true)
}
